mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergConfigTextParser.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive;
+
+import java.util.Locale;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+class HiveIcebergConfigTextParser {
+  private static final int PARTITIONING_TEXT_MAX_LENGTH = 1000;
+
+  private HiveIcebergConfigTextParser() {
+  }
+
+  public static PartitionSpec fromPartitioningText(Schema schema, String text) {
+    Preconditions.checkArgument(text.length() < PARTITIONING_TEXT_MAX_LENGTH,
+        "Partition spec text too long: max allowed length %s, but got %s",
+        PARTITIONING_TEXT_MAX_LENGTH, text.length());
+    PartitionSpec.Builder builder = PartitionSpec.builderFor(schema);
+
+    StringBuilder sb = new StringBuilder();
+    boolean inTransform = false;
+    boolean inEscape = false;
+    boolean needBuild = false;
+    String transformPart = null;
+    String[] argParts = {null, null};
+    int argIndex = 0;
+    for (int i = 0; i < text.length(); i++) {
+      char curr = text.charAt(i);
+      if (needBuild) {
+        Preconditions.checkArgument(curr == ' ' || curr == '|',
+            "Cannot parse partitioning text: end of bracket not followed by ' ' or '|' at position %s in %s", i, text);
+      }
+
+      switch (curr) {
+        case '(':
+          Preconditions.checkArgument(!inTransform,
+              "Cannot parse partitioning text: found multiple ( at position %s in %s", i, text);
+          inTransform = true;
+          transformPart = sb.toString();
+          sb.delete(0, sb.length());
+          break;
+        case ')':
```

Comment:
We might want to do some sanity check for ')' if we go with char by char parsing, since looks like currently these three cases would pass but ideally they shouldn't: 
- `day(created|employee_info.employer`
- `day(created|employee_info.employer)`
- `day(created|employee_info.employer)|bucket(16,id`